cpp/src/io/parquet/reader_impl.cu

```diff
@@ -694,14 +695,18 @@ table_with_metadata reader::impl::read(int skip_rows, int num_rows, int row_grou
       decode_page_data(chunks, pages, skip_rows, num_rows, chunk_map,
                        out_buffers, stream);
 
-      out_columns.reserve(column_types.size());
       for (size_t i = 0; i < column_types.size(); ++i) {
         out_columns.emplace_back(make_column(column_types[i], num_rows,
                                              out_buffers[i], stream, _mr));
       }
     }
   }
 
+  // Create empty columns as needed
+  for (size_t i = out_columns.size(); i < column_types.size(); ++i) {
```

Comment:
Do the empty columns always go last?

Reply:
At that point, out_columns.size() can only be zero or equal to column_types.size(), so the loop will either fill-in all columns or none.

Reply:
Then this doesn't need to be a for loop, it can be an algorithm.

Reply:
Or it could be a simple for loop (went with that) ~ I try to not be idealistic about such things :)

Reply:
This actually is a loop because of `emplace_back`. By contrast with `push_back`, for which one can use `std::transform()` with iterator adapter `std::back_inserter`, there is no equivalent for `emplace_back`. AFAIK, there's no `back_emplacer`. One can write a `back_emplacer` but it's quite a lot of boilerplate code.

Reply:
FWIW, the [related functionality in ORC](https://github.com/rapidsai/cudf/blob/branch-0.13/cpp/src/io/orc/reader_impl.cu#L638) ended up with
```
if (num_rows <= 0 || selected_stripes.size() == 0) {
    std::transform(column_types.cbegin(), column_types.cend(),
                   std::back_inserter(out_columns),
                   [](auto const& dtype) { return make_empty_column(dtype); });
}
```
Personally, I find the for loop more readable.

Reply:
> FWIW, the [related functionality in ORC](https://github.com/rapidsai/cudf/blob/branch-0.13/cpp/src/io/orc/reader_impl.cu#L638) ended up with
> 
> ```
> if (num_rows <= 0 || selected_stripes.size() == 0) {
>     std::transform(column_types.cbegin(), column_types.cend(),
>                    std::back_inserter(out_columns),
>                    [](auto const& dtype) { return make_empty_column(dtype); });
> }
> ```

Yes, but `emplace_back` is (typically) more efficient. And the snippet quoted here ends up using `push_back` (because of the `back_inserter`) which is less efficient than `emplace_back` because it copies, rather than moves. So, unless I'm missing something, using a loop is actually more efficient in this case, because of `emplace_back` (which, again, is not amenable to `transform`, yet).

As far as readability...I think it's a matter of taste. And when it boils down to just matters of taste, as a reviewer, I prefer to respect the OP's tastes rather than impose mine ;-)